sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,243 @@
+            _requestConditions = requestConditions;
+            _length = -1;
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadInternal(
+                buffer,
+                offset,
+                count,
+                async: false,
+                default)
+            .EnsureCompleted();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => await ReadInternal(
+                buffer,
+                offset,
+                count,
+                async: true,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        public async Task<int> ReadInternal(byte[] buffer, int offset, int count, bool async, CancellationToken cancellationToken)
+        {
+            ValidateReadParameters(buffer, offset, count);
+
+            if (_position == _length)
+            {
+                return 0;
+            }
+
+            if (_bufferPosition == 0 || _bufferPosition == _buffer.Length)
+            {
+                int lastDownloadedBytes = await DownloadInternal(async, cancellationToken).ConfigureAwait(false);
+                if (lastDownloadedBytes == 0)
+                {
+                    return 0;
+                }
+            }
+
+            int remainingBytesInBuffer = _bufferLength - _bufferPosition;
+
+            // We will return the minimum of remainingBytesInBuffer and the count provided by the user
+            int bytesToWrite = Math.Min(remainingBytesInBuffer, count);
+
+            Array.Copy(_buffer, _bufferPosition, buffer, offset, bytesToWrite);
+
+            _position += bytesToWrite;
+
+            return bytesToWrite;
+        }
+
+        private async Task<int> DownloadInternal(bool async, CancellationToken cancellationToken)
+        {
+            Response<IDownloadedContent> response;
+
+            HttpRange range = new HttpRange(_position, _bufferSize);
+
+#pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            response = await _downloadInternalFunc(range, _requestConditions, default, async, cancellationToken).ConfigureAwait(false);
+#pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+
+            using Stream networkStream = response.Value.Content;
+
+            int copiedBytes;
+
+            if (async)
+            {
+                copiedBytes = await networkStream.ReadAsync(
+                    buffer: _buffer,
+                    offset: 0,
+                    count: _buffer.Length,
+                    cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                copiedBytes = networkStream.Read(
+                    buffer: _buffer,
+                    offset: 0,
+                    count: _buffer.Length);
+            }
+
+            networkStream.Dispose();
```

Comment:
using/finally?

Reply:
So we can remove `networkStream.Dispose();` because of the using above?  You want me to add a finally block?

Reply:
We need to decide on one option, if we are fine to keep the stream active until the end `using var` is enough, if we want to explicitly dispose at this `using` or `try\finally` should be used.

Having both is a bit confusing.

Reply:
removed `networkStream.Dispose();`